frontend/src/views/instructor/ddahs/import.tsx

```diff
@@ -0,0 +1,319 @@
+import React from "react";
+import { Alert } from "react-bootstrap";
+import { FaArrowRight } from "react-icons/fa";
+import { useSelector } from "react-redux";
+import { assignmentsSelector, applicantsSelector } from "../../../api/actions";
+import { ddahsSelector, upsertDdahs } from "../../../api/actions/ddahs";
+import {
+    Ddah,
+    Assignment,
+    Applicant,
+    MinimalDdah,
+    RawDuty,
+} from "../../../api/defs/types";
+import { ImportActionButton } from "../../../components/import-button";
+import { DiffSpec, diffImport, getChanged } from "../../../libs/diffs";
+import { normalizeDdahImports } from "../../../libs/import-export";
+import { useThunkDispatch } from "../../../libs/thunk-dispatch";
+
+enum DiffType {
+    Unchanged = "UNCHANGED",
+    Created = "CREATED",
+    Updated = "UPDATED",
+    Deleted = "DELETED",
+}
+
+type MinimalDuty = {
+    description: string;
+    hours: number;
+};
+
+interface DdahUpdate {
+    updatedDdah: Ddah;
+    oldDdah: Ddah;
+}
+
+interface DutyDiff {
+    oldDuty?: MinimalDuty;
+    newDuty?: MinimalDuty;
+    status: DiffType;
+}
+
+function getDutiesDiff(oldDuties: RawDuty[], newDuties: RawDuty[]) {
+    let changes: DutyDiff[] = [];
+    let dutiesChanged = false;
+
+    oldDuties.forEach((oldDuty) => {
+        let oldDutyDeleted = true;
+
+        newDuties.forEach((newDuty, index) => {
+            if (
+                oldDuty.description === newDuty.description &&
+                oldDuty.hours === newDuty.hours
+            ) {
+                changes.push({ oldDuty, newDuty, status: DiffType.Unchanged });
+                newDuties.splice(index, 1);
+                oldDutyDeleted = false;
+            } else if (
+                oldDuty.description === newDuty.description &&
+                oldDuty.hours !== newDuty.hours
+            ) {
+                changes.push({ oldDuty, newDuty, status: DiffType.Updated });
+                newDuties.splice(index, 1);
+                oldDutyDeleted = false;
+                dutiesChanged = true;
+            }
+        });
+
+        // If we did not find old duties to match the new one - mark it as created
+        if (oldDutyDeleted) {
+            changes.push({ oldDuty, status: DiffType.Deleted });
+            dutiesChanged = true;
+        }
+    });
+
+    // The new duties that are left were just created - add them
+    newDuties.forEach((duty) =>
+        changes.push({ newDuty: duty, status: DiffType.Created })
+    );
```

Comment:
`changes = changes.concat(newDuties.map(...))` or `changes.pusht(...newDuties.map(...))` if you're a fan of mutability.

Reply:
I do prefer the `forEach` here, is there a reason why you'd prefer `map`? I understand it's more functional-like style, but out of these two:

```
changes = changes.concat(newDuties.map(duty => ({ newDuty: duty, status: DiffType.Created })))

newDuties.forEach((duty) =>
        changes.push({ newDuty: duty, status: DiffType.Created })
    );
```

The second one reads better to me. Let me know what you think!

Reply:
Well, you can do https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap if you prefer, but I like to avoid mutability where I can.